File: instructor Henry/lenguajes/Ruby/clases abstractas/clase abstracta.rb
# Clase abstacta en Ruby

#Ruby no tiene una palabra clave específica para definir clases abstractas, pero puedes simular una clase abstracta utilizando
#raise en el método que debería ser implementado por las subclases.

class Animal
    def emitir_sonido
      raise NotImplementedError, 'se debe implementar este metodo'
    end
  
    def dormir
      puts "Zzz"
    end
  end
  
  class Dog < Animal
    def emitir_sonido
      puts "Woof!"
    end
  end
  
  dog = Dog.new
  dog.emitir_sonido  # Woof!
  dog.dormir       # Zzz
  

File: instructor Henry/lenguajes/Dart/clases abstractas/clase abstracta.dart
## Clases abstractas en Dart

//En Dart, las clases abstractas se declaran con la palabra clave abstract. Un método abstracto no tiene cuerpo.

abstract class Animal {
  void emitirSonido(); // Método abstracto

  void dormir() {
    print("Zzz");
  }
}

class Dog extends Animal {
  void emitirSonido() {
    print("Woof!");
  }
}

void main() {
  Dog dog = Dog();
  dog.emitirSonido(); // Woof!
  dog.dormir();     // Zzz
}
